### ISSUE TITLE
No way to change monster name (or any other)

### ISSUE THREAD

**Issue by bitcraft** [Tuxemon/Tuxemon#156]
Monsters and Trainers have the ability to have different names, currently.  So, it is technically possible to have a Rockitten named "Floofie", and Big Fin named "Tuna".  Right now, there isn't a way to actually change their names though.  So we need a state/menu to input text.

Right now, I see two needs for text input: Naming things and inputing IP/hostnames for multiplayer.  Here is a proposal to create just the Name Input state.
- New Menu/State
- Includes portrait of monster/trainer to be named
- Keyboard input
- Mouse/touch input
- Controller/Joystick input

This state should behave like a tablet computer/phone that we are all familiar with as well as a console where you press the arrows to move a cursor and select letters and numbers.

Here's a rough outline of how to make it:
- Subclass the Menu State
- Overload Menu.initialize_items: create menu items for each letter and symbol
- Overload Menu.on_menu_selection: build input string from selections

Thats the 10,000ft. overview.  Any takers?

**Comment by buddja** [comment]
I have been working on a way to get a string from the user. I found extext.py in _core/components_ but it seems keyboard based from what i've seen (I havent played much with it though) witch wouldn't be what we wanted, so i'm trying to create a onscreen keyboard that the user can navigate with the arrow keys... 

Am i on the right path or did i miss something?

**Comment by bitcraft** [comment]
Glad you've volunteered  :D. I've gone through eztext.py and the update() function would be the most useful part to keep.  Most of the rest will have to be written again to support the new menus and also be touch/joystick/keyboard aware.

Here's a few pointers to get you started:

- overload Menu.handle_event() to grab keypresses
- reuse the update code to enable keyboard input
- for just testing and debugging, store the input string as whatever you like
- for joystick and touch, create many buttons in Menu.initialize_items()
- the buttons should have their callback set to a function to add their key to the stored string
- To display the text, create a TextArea in the menu, and use Menu.alert() to show your text

If you follow those points, you will be 90% complete, with the following problems:

- The menu will be able to input text but each keypress or button will make the text box input clear all the text and draw it again.
- The backspace button will not work
- English letters only

If you get to that point, even with that problem, I will help complete it.  There will be a few issues with showing the input and also supporting other languages besides English.  If you get it mst of the way, I can help finish it.

**Comment by buddja** [comment]
Do you think it should be implemented as a new state something like KeyboardState, or just as another game component that states can call?

 I'm unsure witch one would be better.

**Comment by bitcraft** [comment]
Thats a good point.  Right now, there isn't really a sold "widget system" for directing input, etc.  States are basically the building block right now.  So, I would just treat everything as a "Player/Monster Name State", and when the times comes, we can pull code for a "keyboard widget" or something like that.

TL;DR just make it a state and don't worry too much about code reuse right now.

**Comment by bitcraft** [comment]
Hey buddja, have you been working on this at all?

**Comment by bitcraft** [comment]
I've done some work on this.  Fairly featured, here the state of it:

- Able to use keys to move cursor and select letters
- Able to backspace
- Works with keyboard
- Works with mouse and touch

Just some opinions here...  Do we include capital letters?  I'm linking some game screen shots.  But it kinda boils down to a few options:

- No capitials
- Capitals
- Symbols

This doesn't address locale differences.  For now, it will be English alphabet only, but adding other languages won't be too difficult.

So far, it adds touch/mouse support.  Its ugly now, I know, but a background will make it look fine.

Current issues blocking this:

- "ENTER", "BACKSPACE" buttons render in the grid, and cannot be seen

Work in progress!
![screen shot 2016-08-05 at 2 22 48 pm](https://cloud.githubusercontent.com/assets/556711/17448118/2ac7ab7a-5b18-11e6-91de-c9f89d34b58c.png)

Using capitals
![name_registration_ ocarina_of_time](https://cloud.githubusercontent.com/assets/556711/17448084/f3888a3a-5b17-11e6-8841-d9366f6df981.png)

No capitals
![zelda-nathan](https://cloud.githubusercontent.com/assets/556711/17448064/de6f8e00-5b17-11e6-978c-d6f19775b83c.png)

**Comment by bitcraft** [comment]
Input state implemented.  Now we need a way to open it and change names for stuff.  Monsters, for sure.  Should we allow the player to change their name?

**Comment by xirsoi** [comment]
Starting this one next.

**Comment by MirceaKitsune** [comment]
@xirsoi cheers. My first thought on this issue was that we already have a text input for naming, used when you start a new game and are asked for the player's name. The easiest route seems like calling the existing name editor, possibly straight from the [npc.py add_monster](https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/core/npc.py#L532) function, with the difference that it appends the inputted string to `monster.name` instead.

**Comment by xirsoi** [comment]
My thoughts as well, but I'm also going to add "player_rename" and "rename_monster" script commands. I think the "rename_monster" will have to pull up a version of the party menu so you can select the monster to rename.

**Comment by xirsoi** [comment]
Could you close this @bitcraft ?

**Comment by Sanglorian** [comment]
Nice job, @xirsoi! Also - there's a $50 bounty on this issue! If you claim it on Bountysource then I think bitcraft can approve it and you'll get the money: https://www.bountysource.com/issues/35801375-no-way-to-change-monster-name-or-any-other

**Comment by xirsoi** [comment]
bitcraft did the hardest part of creating the virtual keyboard, i don't feel entirely comfortable claiming the $50.

**Comment by bitcraft** [comment]
oh dang i remember that.  its been up for a long time.  we can discuss the bounty later.  i do recall doing the input state in order to claim it.  :smiling_imp:

**Comment by xirsoi** [comment]
This issue should be closed.
